File: Week1/Union-Find/InterviewQuestions/C++/SuccessorWithDelete/src/main/C++/SuccessorWithDelete.cpp
// swd
#include "SuccessorWithDelete.h"

// Boost
#include <boost/lexical_cast.hpp>

// C++
#include <stdexcept>
#include <string>

namespace swd
{
  SuccessorWithDelete::SuccessorWithDelete(size_type N)
    : m_N(N)
    , m_uf(N)
    , m_removed(N, false)
  {}

  void SuccessorWithDelete::remove(size_type x)
  {
    this->check(x);

    if (!this->m_removed[x])
    {
      this->m_removed[x] = true;

      if (x > 0)
      {
	// check x-1
	if (this->m_removed[x-1])
	{
	  // x-1 has been removed, join to it.
	  this->m_uf.join(x, x-1);
	}
      }

      if (x < this->m_N-1)
      {
	// check x+1
	if (this->m_removed[x+1])
	{
	  // x+1 has been removed. join to it.
	  this->m_uf.join(x, x+1);
	}
      }
    }
  }

  SuccessorWithDelete::size_type SuccessorWithDelete::successor(size_type x)
  {
    this->check(x);
    
    if (x == this->m_N-1)
    {
      // x is the last element.
      return SuccessorWithDelete::INVALID_INDEX;
    }

    if (!this->m_removed[x])
    {
      // x hasn't been removed.
      return x;
    }
    
    if (this->m_N == 1)
    {
      // Only one element, x, which has been removed.
      return SuccessorWithDelete::INVALID_INDEX;
    }

    // Retrieve largest element of connected component
    // involving removed element, x. The successor, if any,
    // will be the next largest integer.
    const size_type y = 1 + this->m_uf.find(x);

    return y < this->m_N ? y : SuccessorWithDelete::INVALID_INDEX;
  }

  const SuccessorWithDelete::size_type SuccessorWithDelete::INVALID_INDEX = -1;

  void SuccessorWithDelete::check(size_type i) const
  {
    if (i >= this->m_N)
    {
      throw std::out_of_range("Invalid index: " + boost::lexical_cast<std::string>(i));
    }
  }

} // ! namespace swd

File: Week1/Union-Find/InterviewQuestions/Java/SuccessorWithDelete/src/SuccessorWithDelete.java

public class SuccessorWithDelete {

	private boolean[] m_removed;
	private int[] m_max;
	private WeightedQuickUnionwithPathCompression m_uf;
	private int m_N;
	
	private void validate(int x) {
		if (x < 0 || x >= this.m_N) {
			throw new IllegalArgumentException("Index " + x + " is outside of valid range [0, " + this.m_N + ").");
		}
	}
	
	public SuccessorWithDelete(int N) {
		
		if (N < 1) {
			throw new IllegalArgumentException("Invalid number of elements.");
		}
		
		this.m_removed = new boolean[N];
		this.m_max = new int[N];
		this.m_uf = new WeightedQuickUnionwithPathCompression(N);
		this.m_N = N;
		
		for (int i = 0; i < N; ++i) {
			this.m_max[i] = i;
		}
	}
	
	public void remove(int x) {
		this.validate(x);
		
		// Update removed flag.
		this.m_removed[x] = true;
		
		if (x > 0) {
			// check if x - 1 is removed.
			if (this.m_removed[x-1]) {
				// join x to x-1's connected component.
				this.m_uf.join(x-1, x);
				
				// x is now the largest in its connected component.
				this.m_max[this.m_uf.root(x)] = x;
			}
		}
		
		if (x < this.m_N-1) {
			// check if x+1 is removed.
			if (this.m_removed[x+1]) {

				// Get max of adjacent component.
				final int adjacentMax = this.m_max[this.m_uf.root(x+1)];
				
				// join x to x+1's connected component.
				this.m_uf.join(x, x+1);
				
				// Set max of x's component to adjacent max.
				this.m_max[this.m_uf.root(x)] = adjacentMax;
			}
		}
	}
	
	public int successor(int x) {
		this.validate(x);
		
		if (x == this.m_N-1) {
			return -1;
		}
		
		if (!this.m_removed[x+1]) {
			return x+1;
		}
		
		final int adjacentRoot = this.m_uf.root(x+1);
		int adjacentMax = this.m_max[adjacentRoot];
		
		final int suc = 1+adjacentMax;
		return suc < this.m_N ? suc : -1;
	}
}
